File: homework3/q5_First_K_BinaryNUmbers.py
# Time complexity: O(k)
# Space complexity: O(k)
#Time Spent: 28 minutes

from collections import deque

def FirstKBinaryNumbers(k):
    if k < 1:
        return None
    
    res = []
    q = deque()

    q.append("1")

    while k > 0:
        curr = q.popleft()
        res.append(curr)

        q.append(curr + "0")
        q.append(curr + "1")

        k -= 1

    return res


if __name__ == "__main__":
    ks = (1, 2, 3, 4, 5)
    for k in ks:
        print(f"K is {k}:")
        print(FirstKBinaryNumbers(k))

File: Homework3_Rodrigo_Canto/q5_FirstKBinaryNumbers.cpp
//Time Complexity: O(K * log(k)).
//Space Complexity: O(K * log(k)).
//Tecnhique: Math.

#include "bits/stdc++.h"
using namespace std;

vector<string> solveFirstKBinaryNumbers(int k){

    vector<string> answer;

    for(int i = 0; i < k; ++i){

        int number = i;

        string binary_representation = "";

        //Handle the special case (zero).
        if(number == 0){
            binary_representation = "0";
        }
        else{


            while(number != 0){

                //The AND operation allows us to process the last bit
                if(number & 1){
                    binary_representation.push_back('1');
                }
                else{
                    binary_representation.push_back('0');
                }

                //The right shift "eliminates" the last bit we just processed.
                number = number / 2;
            }
        }

        //We process the bits from right to left so we need to reverse the string.
        reverse(binary_representation.begin(), binary_representation.end());

        answer.push_back(binary_representation);
    }

    return answer;
}

int main() {

    int k; cin >> k;

    for(string binary_number : solveFirstKBinaryNumbers(k)){
        cout << binary_number << " ";
    }
}

//Time Spent: 10 minutes